type_resolver.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2023-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vanguard
+
+import (
+	"google.golang.org/protobuf/reflect/protoreflect"
+	"google.golang.org/protobuf/reflect/protoregistry"
+	"google.golang.org/protobuf/types/dynamicpb"
+)
+
+// TypeResolver can resolve message and extension types and is used to instantiate
+// messages as needed for the middleware to serialize/de-serialize request and
+// response payloads.
+//
+// Implementations of this interface should be comparable, so they can be used as
+// map keys. Typical implementations are pointers to structs, which are suitable.
+type TypeResolver interface {
+	protoregistry.MessageTypeResolver
+	protoregistry.ExtensionTypeResolver
+}
+
+type fallbackResolver []TypeResolver
```

Comment:
I think this `fallbackResolver` is not needed. At the time of register we have the message descriptor and can use it to create the dynamic type. Why do we need to create the resolver?

Reply:
I mentioned this in a [comment above](https://github.com/connectrpc/vanguard-go/pull/117#discussion_r1478517834) and also put it in a [comment in the code](https://github.com/connectrpc/vanguard-go/commit/bad91a523a2e52f4577a33ab90edc25b005bcb9e#diff-cd37f93851046afde3cffe0e07e2a9181545e252442ec5378b08919ce254fcdcR113-R116).

The resolver is used for more than just the message types for requests and responses. It is also used to resolve other message types (like inside `Any` values) and extensions.

Reply:
The behaviour isn't consistent with `Any` values because if the service does pass the `canUseGlobaTypes` check the dynamic types won't be pulled in for the any values: https://github.com/connectrpc/vanguard-go/pull/117/files#diff-cd37f93851046afde3cffe0e07e2a9181545e252442ec5378b08919ce254fcdcR97-R99

Reply:
If `canUseGlobalTypes` returns true, there are no dynamic types needed - the file (and thus its entire dependency graph, including all messages and extensions) is already registered in `protoregistry.GlobalTypes`.

Reply:
I guess we could strengthen `canUseGlobalTypes` by having it also verify that the resolve message types are _not_ dynamic. That would eliminate the possible corner case where someone registers dynamic descriptors and types into `protoregistry.GlobalFiles` and `protoregistry.GlobalTypes`, but fails to do so thoroughly (i.e. doesn't register all message and extension types or all files in the transitive dependency graph).

But it seemed like overkill to go that far. (And if someone does abuse the global registries that way, and it causes some issues in their vanguard-powered server, they can work around it by specifying a custom resolver using the `WithTypeResolver` option.)

Reply:
ping

Reply:
I think its fine as is. I'm not sure the Any benefit is that useful over forcing the user to provide their own resolver. The dynamicpb changes are a good user improvement.